packages/truffle-decode-utils/src/types/types.ts

```diff
@@ -0,0 +1,541 @@
+  }
+
+  export interface ArrayTypeDynamic {
+    typeClass: "array";
+    kind: "dynamic";
+    baseType: Type;
+    location?: string;
+  }
+
+  export type ElementaryType = UintType | IntType | BoolType | BytesType | FixedType
+    | UfixedType | AddressType | StringType;
+
+  export interface MappingType {
+    typeClass: "mapping";
+    keyType: ElementaryType;
+    valueType: Type;
+    location?: string; //should only ever be storage
+  }
+
+  export interface FunctionType {
+    typeClass: "function";
+    visibility: string; //should be "internal" or "external", not weird
+    //solidity-internal visibilities (but for technical reasons I'm not
+    //putting that in the type requirements)
+    mutability: string; //similarly, should be "nonpayable", "pure", "view",
+    //or "payable"
+    inputParameterTypes: Type[];
+    outputParameterTypes: Type[];
+    //we do not presently support bound functions
+  }
+
+  export type ContractDefinedType = StructType | EnumType;
+  export type UserDefinedType = ContractDefinedType | ContractType;
+
+  export interface StructType {
+    typeClass: "struct";
+    id: number;
+    typeName: string;
+    definingContractName: string;
+    definingContract?: ContractType;
+    memberTypes?: [string, Type][];
+    location?: string;
+  }
+
+  export interface EnumType {
+    typeClass: "enum";
+    id: number;
+    typeName: string;
+    definingContractName: string;
+    definingContract?: ContractType;
+    options?: string[]; //these should be in order
+  }
+
+  export interface ContractType {
+    typeClass: "contract";
+    id: number;
+    typeName: string;
+    contractKind?: string;
+    payable?: boolean; //will be useful in the future
+    //may have more optional members defined later, but I'll leave these out for
+    //now
+  }
+
+  export interface MagicType {
+    typeClass: "magic";
+    variable: string; //not putting this in the type annotation for technical
+    //reasons, but this should be one of "message", "block", or "transaction";
+    //we do *not* presently support abi or meta_type
+    memberTypes?: {
+      [field: string]: Type
+    };
+  }
+
+  export type ReferenceType = ArrayType | MappingType | StructType | StringType | BytesTypeDynamic;
+
+  export interface TypesById {
+    [id: number]: UserDefinedType;
+  };
+
+  //NOTE: the following function will *not* work for arbitrary nodes! It will,
+  //however, work for the ones we need (i.e., variable definitions, and arbitrary
+  //things of elementary type)
+  //NOTE: set forceLocation to *null* to force no location. leave it undefined
+  //to not force a location.
+  //NOTE: set compiler to null to force addresses to *not* be payable (HACK)?
+  export function definitionToType(definition: AstDefinition, compiler: Contexts.CompilerVersion | null, forceLocation?: string | null): Type {
+    debug("definition %O", definition);
+    let typeClass = DefinitionUtils.typeClass(definition);
+    switch(typeClass) {
```

Comment:
Huh... so basically, switch on the typeclass and then dispatch to a different function based on the result, rather than inlining?  I mean, I guess?  I don't see that factoring that really gains us anything.  My own preference would be for leaving it as-is.

Btw note that I was planning on adding `abiParameterToType` in the future as well.

Reply:
What I'm proposing is to go a step further than switching on the typeclass... well depending on what you mean.

What I mean is something like:
```typescript
type TypeConstructorMap = { [typeClass: string]: (/* ...args? */) => Types.Type }
```

Then you don't need a switch statement at all, but instead just do an object lookup.

Reply:
Oh, I see.  Yeah, that just seems slightly more opaque and not essentially different?

Reply:
Also as mentioned to account for the future `abiParameterToType` this would need to be, like, `DefinitionTypeConstructorMap` so you could also have an `AbiTypeConstructorMap` or whatever.

Reply:
OK, seems we've decided to leave this alone for now.